parallel.go

```diff
@@ -0,0 +1,221 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package parallel implements job processing using parallel workers.
+package parallel
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/api/iterator"
+)
+
+type contextKey int
+
+const (
+	parallelKey contextKey = iota
+)
+
+// TestSerialize forces a Parallel runner to run jobs serially and in the strict
+// FIFO order.
+func TestSerialize(ctx context.Context) context.Context {
+	return context.WithValue(ctx, parallelKey, true)
+}
+
+func isSerialized(ctx context.Context) bool {
+	v, ok := ctx.Value(parallelKey).(bool)
+	return ok && v
+}
+
+// Job is a unit of work to be done, which returns a result and an error.
+type Job = func() (interface{}, error)
+
+// JobsIter is a generator of work units for executing in Map (the API).  The
+// iterator loosely complies with Google Cloud iterator guidelines: it either
+// returns a work unit as a callable function, or an error. When the error is
+// iterator.Done, it's the end of the sequence (and not really an error).
+type JobsIter interface {
+	Next() (Job, error)
+}
+
+// ResultsIter is an iterator over results, typically merged from parallel runs.
+type ResultsIter interface {
+	Next() (interface{}, error)
+}
+
+// Result of a Job as a single value, including the error.
+type Result struct {
+	Value interface{}
+	Error error
+}
+
+// Map runs multiple jobs in parallel on a given number of workers, 0=unlimited,
+// collects their results and returns in an iterator. The order of results is
+// not guaranteed, unless the number of workers is 1.
+//
+// Canceling the supplied context, or calling Stop() immediately stops queuing
+// new jobs, but the jobs that already started will finish and their results
+// will be returned.
+//
+// Example usage:
+//
+// m := NewMap(ctx, 2, jobsIter)
+// for {
+//   v, err := m.Next()
+//   if err == iterator.Done {
+//     break
+//   }
+//   // Process v and err.
+// }
+//
+// Note, that Next() may return a non-nil error from a job, but only
+// iterator.Done means all the jobs have finished.
+type Map struct {
+	ctx     context.Context // derived cancelable context
+	cancel  func()          // function that cancels ctx
+	workers int             // maximum number of parallel jobs allowed
+	jobs    int             // number of jobs currently running
+	it      JobsIter        // job iterator
+	resCh   chan Result     // workers send their results to this channel
+	done    error           // non-nil error returned by the job iterator
+	mux     sync.Mutex      // to make Next() go routine safe
+}
+
+var _ ResultsIter = &Map{}
+
+// NewMap creates a new parallel map runner. If workers <= 0, the number
+// of workers is unlimited.
+func NewMap(ctx context.Context, workers int, it JobsIter) *Map {
```

Comment:
Rename this to `Map`, and make the original `Map` an internal struct. There is no need to expose the implementation type.